pkg/lib/machine-dialogs.js

```diff
@@ -739,76 +900,85 @@ function ChangeAuth(dialog) {
                 template = "auth-failed";
 
             locked_identity = error_options.error && error_options.error.startsWith("locked identity");
+
+            offer_login_password = !!available.password;
+            offer_key_password = locked_identity;
+        } else {
+            offer_login_password = true;
+            offer_key_password = false;
         }
 
         dialog.render({
-            supported : methods,
-            available : available,
-            machine_user : machine_user,
-            default_user : self.user ? self.user.name : "",
-            show_password : available && available.password && !no_password,
-            show_ticket: available && available['gssapi-mic'] && have_ticket,
-            show_locked_identity: available && available['public-key'] && locked_identity,
-            can_sync: !!dialog.codes['sync-users'],
-            'machines.allow_connection_string' : machines.allow_connection_string,
-            sync_link : function() {
-                return function(text, render) {
-                    return '<a tabindex="0" id="do-sync-users">' + render(text) + "</a>";
-                };
-            }
+            available : offer_login_password || offer_key_password,
+            only_password: offer_login_password && !offer_key_password,
+            only_key: !offer_login_password && offer_key_password,
+            password_and_key: offer_login_password && offer_key_password,
+            key: identity_path
         }, template);
 
         if (methods === null && machines.has_auth_results) {
             promise = dialog.try_to_connect(dialog.address)
                     .fail(function(ex) {
-                        error_options = ex;
-                        render();
+                        if (ex.problem && dialog.codes[ex.problem] != "change-auth") {
+                            dialog.render_error(ex);
+                        } else {
+                            set_error_options(ex);
+                            render();
+                        }
                     })
 
             // Fixed already, just close
                     .done(function (v) {
-                        dialog.complete(v);
+                        if (machine) {
```

Comment:
Is this still relevant? This mattered when ChangeAuth would silently switch to a different user (so that AddHost would fail, but ChangeAuth would immediately succeed), but that probably doesn't happen anymore.